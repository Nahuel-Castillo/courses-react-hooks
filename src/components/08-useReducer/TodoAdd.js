import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description:inputValue }, handleInputChange, reset ] = useForm( { description: '' });

    const handleSubmit = e => {
        e.preventDefault();

        if ( inputValue.trim().length <= 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: inputValue,
            done: false
        };

        handleAddTodo(newTodo);

        reset();
    }

    return (
        <>
            <h4> Add TODO </h4>
            <hr/>

            <form onSubmit={ handleSubmit }>

                <input 
                    type="text"
                    name="description"
                    placeholder="Learn..."
                    autoComplete="off"
                    className="form-control"
                    onChange={ handleInputChange }
                    value={ inputValue }
                />

                <button
                    type="submit"
                    className="btn btn-outline-success mt-1 btn-block"
                >
                    add
                </button>

            </form>
        </>
    )
}
