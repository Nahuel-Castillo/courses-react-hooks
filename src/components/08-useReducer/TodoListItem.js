
import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ( { todo, i, handleDelete, handleToggle } ) => {
    return (
        <li
            className="list-group-item"
        >
            <p
                className={`${todo.done && 'complete'}`}
                onClick={ () => handleToggle(todo.id) }
            >
                {i + 1}. {todo.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                Remove
            </button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    i: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}