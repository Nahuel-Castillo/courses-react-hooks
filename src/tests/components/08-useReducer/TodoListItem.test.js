import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test of <TodoListItem/>', () => {
    
    const todo = demoTodos[0];
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    
    let wrapper = shallow(
        <TodoListItem
            todo={ todo }
            i={ todo.id }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />
    );
    
    test('should show corretly', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should call handleDelete()', () => {
        
        wrapper.find('button').simulate('click');

        expect( handleDelete ).toHaveBeenCalled();
    });

    test('should call handleToggle()', () => {
        
        wrapper.find('p').simulate('click');

        expect( handleToggle ).toHaveBeenCalledWith( todo.id );

    });

    test('should show text correctly', () => {
        
        const p = wrapper.find('p');

        expect( p.text().trim() ).toBe( `${todo.id + 1}. ${todo.desc}` );
    });
});