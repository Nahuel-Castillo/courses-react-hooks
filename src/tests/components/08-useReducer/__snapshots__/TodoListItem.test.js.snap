// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <TodoListItem/> should show corretly 1`] = `
<li
  className="list-group-item"
>
  <p
    className="complete"
    onClick={[Function]}
  >
    2
    . 
    Learn React
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Remove
  </button>
</li>
`;
